### ISSUE TITLE
PGPException occurred when signed commit verification

### ISSUE THREAD

**Issue by ray45422** [gitbucket/gitbucket#2495]
## Issue
**Impacted version**: 4.34.0, 4.33.0

**Deployment mode**: standalone app, with nginx

**Problem description**:
An error occurred when verifying the commit signed by EDDSA key.
Updating to bouncycastle 1.66 seems to work.
```
org.bouncycastle.openpgp.PGPException: cannot recognise keyAlgorithm: 22
	at org.bouncycastle.openpgp.operator.bc.BcImplProvider.createSigner(Unknown Source)
	at org.bouncycastle.openpgp.operator.bc.BcPGPContentVerifierBuilderProvider$BcPGPContentVerifierBuilder.build(Unknown Source)
	at org.bouncycastle.openpgp.PGPSignature.init(Unknown Source)
	at gitbucket.core.util.GpgUtil$.$anonfun$verifySign$3(GpgUtil.scala:48)
	at scala.Option.map(Option.scala:242)
	at gitbucket.core.util.GpgUtil$.$anonfun$verifySign$2(GpgUtil.scala:47)
	at scala.collection.Iterator$$anon$10.nextCur(Iterator.scala:583)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:597)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:589)
	at scala.collection.mutable.ListBuffer.addAll(ListBuffer.scala:118)
	at scala.collection.mutable.ListBuffer$.from(ListBuffer.scala:38)
	at scala.collection.immutable.List$.from(List.scala:617)
	at scala.collection.IterableOnceOps.toList(IterableOnce.scala:1234)
	at scala.collection.IterableOnceOps.toList$(IterableOnce.scala:1234)
	at scala.collection.AbstractIterator.toList(Iterator.scala:1274)
	at gitbucket.core.util.GpgUtil$.verifySign(GpgUtil.scala:40)
	at gitbucket.core.controller.RepositoryViewerControllerBase.$anonfun$$init$$19(RepositoryViewerController.scala:296)
	at scala.Option.flatMap(Option.scala:283)
	at gitbucket.core.controller.RepositoryViewerControllerBase.$anonfun$$init$$18(RepositoryViewerController.scala:295)
	at scala.collection.immutable.List.map(List.scala:219)
	at gitbucket.core.controller.RepositoryViewerControllerBase.$anonfun$$init$$17(RepositoryViewerController.scala:281)
	at scala.util.Using$.resource(Using.scala:252)
	at gitbucket.core.controller.RepositoryViewerControllerBase.$anonfun$$init$$12(RepositoryViewerController.scala:267)
	at gitbucket.core.util.ReferrerAuthenticator.$anonfun$authenticate$4(Authenticator.scala:95)
	at scala.Option.map(Option.scala:242)
	at gitbucket.core.util.ReferrerAuthenticator.authenticate(Authenticator.scala:93)
	at gitbucket.core.util.ReferrerAuthenticator.referrersOnly(Authenticator.scala:87)
	at gitbucket.core.util.ReferrerAuthenticator.referrersOnly$(Authenticator.scala:87)
	at gitbucket.core.controller.RepositoryViewerController.referrersOnly(RepositoryViewerController.scala:42)
	at gitbucket.core.controller.RepositoryViewerControllerBase.$anonfun$$init$$11(RepositoryViewerController.scala:252)
	at org.scalatra.ScalatraBase.liftAction(ScalatraBase.scala:282)
	at org.scalatra.ScalatraBase.$anonfun$invoke$1(ScalatraBase.scala:276)
	at org.scalatra.ApiFormats.withRouteMultiParams(ApiFormats.scala:185)
	at org.scalatra.ApiFormats.withRouteMultiParams$(ApiFormats.scala:171)
	at gitbucket.core.controller.ControllerBase.withRouteMultiParams(ControllerBase.scala:35)
	at org.scalatra.ScalatraBase.invoke(ScalatraBase.scala:275)
	at org.scalatra.ScalatraBase.invoke$(ScalatraBase.scala:274)
	at gitbucket.core.controller.ControllerBase.invoke(ControllerBase.scala:35)
	at org.scalatra.ScalatraBase.$anonfun$runRoutes$3(ScalatraBase.scala:250)
	at scala.Option.flatMap(Option.scala:283)
	at org.scalatra.ScalatraBase.$anonfun$runRoutes$1(ScalatraBase.scala:248)
	at scala.collection.immutable.Stream.flatMap(Stream.scala:206)
	at org.scalatra.ScalatraBase.runRoutes(ScalatraBase.scala:247)
	at org.scalatra.ScalatraBase.runRoutes$(ScalatraBase.scala:245)
	at gitbucket.core.controller.ControllerBase.runRoutes(ControllerBase.scala:35)
	at org.scalatra.ScalatraBase.runActions$1(ScalatraBase.scala:169)
	at org.scalatra.ScalatraBase.$anonfun$executeRoutes$6(ScalatraBase.scala:181)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at org.scalatra.ScalatraBase.cradleHalt(ScalatraBase.scala:201)
	at org.scalatra.ScalatraBase.executeRoutes(ScalatraBase.scala:181)
	at org.scalatra.ScalatraBase.executeRoutes$(ScalatraBase.scala:150)
	at gitbucket.core.controller.ControllerBase.executeRoutes(ControllerBase.scala:35)
	at org.scalatra.ScalatraBase.$anonfun$handle$1(ScalatraBase.scala:123)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at org.scalatra.DynamicScope.withResponse(DynamicScope.scala:75)
	at org.scalatra.DynamicScope.withResponse$(DynamicScope.scala:73)
	at gitbucket.core.controller.ControllerBase.withResponse(ControllerBase.scala:35)
	at org.scalatra.DynamicScope.$anonfun$withRequestResponse$1(DynamicScope.scala:55)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at org.scalatra.DynamicScope.withRequest(DynamicScope.scala:66)
	at org.scalatra.DynamicScope.withRequest$(DynamicScope.scala:64)
	at gitbucket.core.controller.ControllerBase.withRequest(ControllerBase.scala:35)
	at org.scalatra.DynamicScope.withRequestResponse(DynamicScope.scala:54)
	at org.scalatra.DynamicScope.withRequestResponse$(DynamicScope.scala:52)
	at gitbucket.core.controller.ControllerBase.withRequestResponse(ControllerBase.scala:35)
	at org.scalatra.ScalatraBase.handle(ScalatraBase.scala:123)
	at org.scalatra.ScalatraBase.handle$(ScalatraBase.scala:119)
	at gitbucket.core.controller.ControllerBase.org$scalatra$servlet$ServletBase$$super$handle(ControllerBase.scala:35)
	at org.scalatra.servlet.ServletBase.handle(ServletBase.scala:42)
	at org.scalatra.servlet.ServletBase.handle$(ServletBase.scala:35)
	at gitbucket.core.controller.ControllerBase.org$scalatra$FlashMapSupport$$super$handle(ControllerBase.scala:35)
	at org.scalatra.FlashMapSupport.$anonfun$handle$1(FlashMap.scala:197)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at org.scalatra.DynamicScope.withRequest(DynamicScope.scala:66)
	at org.scalatra.DynamicScope.withRequest$(DynamicScope.scala:64)
	at gitbucket.core.controller.ControllerBase.withRequest(ControllerBase.scala:35)
	at org.scalatra.FlashMapSupport.handle(FlashMap.scala:172)
	at org.scalatra.FlashMapSupport.handle$(FlashMap.scala:171)
	at gitbucket.core.controller.ControllerBase.handle(ControllerBase.scala:35)
	at org.scalatra.ScalatraFilter.$anonfun$doFilter$1(ScalatraFilter.scala:41)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at org.scalatra.ScalatraFilter.doFilter(ScalatraFilter.scala:41)
	at org.scalatra.ScalatraFilter.doFilter$(ScalatraFilter.scala:36)
	at gitbucket.core.controller.ControllerBase.doFilter(ControllerBase.scala:73)
	at gitbucket.core.servlet.CompositeScalatraFilter.$anonfun$process$2(CompositeScalatraFilter.scala:67)
	at gitbucket.core.servlet.CompositeScalatraFilter.$anonfun$process$2$adapted(CompositeScalatraFilter.scala:64)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:553)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:551)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:921)
	at gitbucket.core.servlet.CompositeScalatraFilter.process(CompositeScalatraFilter.scala:64)
	at gitbucket.core.servlet.ControllerFilter.doFilter(CompositeScalatraFilter.scala:25)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618)
	at gitbucket.core.servlet.ControllerFilter.doFilter(CompositeScalatraFilter.scala:31)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618)
	at org.scalatra.ScalatraFilter.$anonfun$doNotFound$1(ScalatraFilter.scala:79)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at org.scalatra.ScalatraBase.$anonfun$executeRoutes$5(ScalatraBase.scala:172)
	at scala.Option.getOrElse(Option.scala:201)
	at org.scalatra.ScalatraBase.$anonfun$executeRoutes$3(ScalatraBase.scala:172)
	at scala.Option.getOrElse(Option.scala:201)
	at org.scalatra.ScalatraBase.runActions$1(ScalatraBase.scala:172)
	at org.scalatra.ScalatraBase.$anonfun$executeRoutes$6(ScalatraBase.scala:181)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at org.scalatra.ScalatraBase.cradleHalt(ScalatraBase.scala:201)
	at org.scalatra.ScalatraBase.executeRoutes(ScalatraBase.scala:181)
	at org.scalatra.ScalatraBase.executeRoutes$(ScalatraBase.scala:150)
	at gitbucket.core.controller.ControllerBase.executeRoutes(ControllerBase.scala:35)
	at org.scalatra.ScalatraBase.$anonfun$handle$1(ScalatraBase.scala:123)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at org.scalatra.DynamicScope.withResponse(DynamicScope.scala:75)
	at org.scalatra.DynamicScope.withResponse$(DynamicScope.scala:73)
	at gitbucket.core.controller.ControllerBase.withResponse(ControllerBase.scala:35)
	at org.scalatra.DynamicScope.$anonfun$withRequestResponse$1(DynamicScope.scala:55)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at org.scalatra.DynamicScope.withRequest(DynamicScope.scala:66)
	at org.scalatra.DynamicScope.withRequest$(DynamicScope.scala:64)
	at gitbucket.core.controller.ControllerBase.withRequest(ControllerBase.scala:35)
	at org.scalatra.DynamicScope.withRequestResponse(DynamicScope.scala:54)
	at org.scalatra.DynamicScope.withRequestResponse$(DynamicScope.scala:52)
	at gitbucket.core.controller.ControllerBase.withRequestResponse(ControllerBase.scala:35)
	at org.scalatra.ScalatraBase.handle(ScalatraBase.scala:123)
	at org.scalatra.ScalatraBase.handle$(ScalatraBase.scala:119)
	at gitbucket.core.controller.ControllerBase.org$scalatra$servlet$ServletBase$$super$handle(ControllerBase.scala:35)
	at org.scalatra.servlet.ServletBase.handle(ServletBase.scala:42)
	at org.scalatra.servlet.ServletBase.handle$(ServletBase.scala:35)
	at gitbucket.core.controller.ControllerBase.org$scalatra$FlashMapSupport$$super$handle(ControllerBase.scala:35)
	at org.scalatra.FlashMapSupport.$anonfun$handle$1(FlashMap.scala:197)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at org.scalatra.DynamicScope.withRequest(DynamicScope.scala:66)
	at org.scalatra.DynamicScope.withRequest$(DynamicScope.scala:64)
	at gitbucket.core.controller.ControllerBase.withRequest(ControllerBase.scala:35)
	at org.scalatra.FlashMapSupport.handle(FlashMap.scala:172)
	at org.scalatra.FlashMapSupport.handle$(FlashMap.scala:171)
	at gitbucket.core.controller.ControllerBase.handle(ControllerBase.scala:35)
	at org.scalatra.ScalatraFilter.$anonfun$doFilter$1(ScalatraFilter.scala:41)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at org.scalatra.ScalatraFilter.doFilter(ScalatraFilter.scala:41)
	at org.scalatra.ScalatraFilter.doFilter$(ScalatraFilter.scala:36)
	at gitbucket.core.controller.ControllerBase.doFilter(ControllerBase.scala:73)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1618)
	at gitbucket.core.servlet.TransactionFilter.$anonfun$doFilter$1(TransactionFilter.scala:39)
	at gitbucket.core.servlet.TransactionFilter.$anonfun$doFilter$1$adapted(TransactionFilter.scala:30)
	at com.github.takezoe.slick.blocking.BlockingJdbcProfile$BlockingAPI$BlockingDatabase.$anonfun$withTransaction$2(BlockingProfile.scala:207)
	at slick.JdbcProfileBlockingSession$BlockingSession.withTransaction(TransactionalJdbcBackend.scala:26)
	at com.github.takezoe.slick.blocking.BlockingJdbcProfile$BlockingAPI$BlockingDatabase.$anonfun$withTransaction$1(BlockingProfile.scala:207)
	at com.github.takezoe.slick.blocking.BlockingJdbcProfile$BlockingAPI$BlockingDatabase.withSession(BlockingProfile.scala:200)
	at com.github.takezoe.slick.blocking.BlockingJdbcProfile$BlockingAPI$BlockingDatabase.withTransaction(BlockingProfile.scala:207)
	at gitbucket.core.servlet.TransactionFilter.doFilter(TransactionFilter.scala:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1610)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:549)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:602)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1610)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1369)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:489)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1580)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1284)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:173)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:501)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:556)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:272)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by gemmaro** [comment]
I encountered the same error.

**Impacted version**: 4.35.3
**Deployment mode**: I installed it on macOS with Homebrew.
**Problem description**: I was trying to view the list of commits and ended up looking at the error page. Here's how it happened.

* First, a user already had a GPG key (I'm not sure if this is a requirement for reproduction).
* I registered an additional GPG key for that user.
* Committed using the additionally registered GPG key.
* When I tried to display the commit history of the repository that contains the commit, I got the same error as above.

If I don't use GPG signing, there is no problem in displaying the history.

**Comment by ray45422** [comment]
This problem occurs when a commit signed by an EDDSA key exists in the commit log and a verifiable GPG key is registered. It will not occur if there is no verifiable key registered.
The cause is that the included version of bouncycastle (1.60) does not support EDDSA.

Reproduction Procedure
1. create a GPG key
```
$ gpg --expert --full-generate-key
gpg (GnuPG) 2.2.27; Copyright (C) 2021 g10 Code GmbH
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Please select what kind of key you want:
   (1) RSA and RSA (default)
   (2) DSA and Elgamal
   (3) DSA (sign only)
   (4) RSA (sign only)
   (7) DSA (set your own capabilities)
   (8) RSA (set your own capabilities)
   (9) ECC and ECC
  (10) ECC (sign only)
  (11) ECC (set your own capabilities)
  (13) Existing key
  (14) Existing key from card
Your selection? 9
Please select which elliptic curve you want:
   (1) Curve 25519
   (3) NIST P-256
   (4) NIST P-384
   (5) NIST P-521
   (6) Brainpool P-256
   (7) Brainpool P-384
   (8) Brainpool P-512
   (9) secp256k1
Your selection? 1
Please specify how long the key should be valid.
         0 = key does not expire
      <n>  = key expires in n days
      <n>w = key expires in n weeks
      <n>m = key expires in n months
      <n>y = key expires in n years
Key is valid for? (0) 0
Key does not expire at all
Is this correct? (y/N) y

GnuPG needs to construct a user ID to identify your key.

Real name: foo bar
Email address: foobar@example.com
Comment:
You selected this USER-ID:
    "foo bar <foobar@example.com>"

Change (N)ame, (C)omment, (E)mail or (O)kay/(Q)uit? O
We need to generate a lot of random bytes. It is a good idea to perform
some other action (type on the keyboard, move the mouse, utilize the
disks) during the prime generation; this gives the random number
generator a better chance to gain enough entropy.
We need to generate a lot of random bytes. It is a good idea to perform
some other action (type on the keyboard, move the mouse, utilize the
disks) during the prime generation; this gives the random number
generator a better chance to gain enough entropy.
gpg: key 273DD0C26CD00B5B marked as ultimately trusted
public and secret key created and signed.

pub   ed25519 2021-05-06 [SC]
      78DAF86F1818BA4795427E91273DD0C26CD00B5B
uid                      foo bar <foobar@example.com>
sub   cv25519 2021-05-06 [E]
```
In this example, 78DAF86F1818BA4795427E91273DD0C26CD00B5B is the key ID.

2. git commit
Make some changes and commit.
The key ID should be changed to the one created in step 1.
`git commit --gpg-sign=78DAF86F1818BA4795427E91273DD0C26CD00B5B`
And push commit.

3. check the commit log
Check the commit log of the repository in gitbucket. At this point, no error will be shown, and Unverified tag will be displayed.

4. register the key
Register your GPG key from your account settings. The following command will output the keys that can be registered.
`gpg --armor --export 78DAF86F1818BA4795427E91273DD0C26CD00B5B`.

4. check the commit log
An error occurs.

**Comment by takezoe** [comment]
A JGit's dependency on the latest version of bouncycastle (1.69) seems to be overridden by apache-sshd 2.1.0's one (1.60) and this old version of bouncycastle doesn't support EDDSA. Probably, this is the root cause of this error.

We have to use newer version of bouncycastle (>= 1.65) to support EDDSA.
